sway/tree/view.c

```diff
@@ -6,94 +7,122 @@
 #include "sway/tree/layout.h"
 #include "sway/tree/view.h"
 
+struct sway_view *view_create(enum sway_view_type type,
+		const struct sway_view_impl *impl) {
+	struct sway_view *view = calloc(1, sizeof(struct sway_view));
+	if (view == NULL) {
+		return NULL;
+	}
+	view->type = type;
+	view->impl = impl;
+	wl_list_init(&view->unmanaged_view_link);
+	return view;
+}
+
+void view_destroy(struct sway_view *view) {
+	if (view == NULL) {
+		return;
+	}
+
+	if (view->surface != NULL) {
+		view_unmap(view);
+	}
+	if (view->swayc != NULL) {
+		container_view_destroy(view->swayc);
+	}
+
+	free(view);
+}
+
 const char *view_get_title(struct sway_view *view) {
-	if (view->iface.get_prop) {
-		return view->iface.get_prop(view, VIEW_PROP_TITLE);
+	if (view->impl->get_prop) {
+		return view->impl->get_prop(view, VIEW_PROP_TITLE);
 	}
 	return NULL;
 }
 
 const char *view_get_app_id(struct sway_view *view) {
-	if (view->iface.get_prop) {
-		return view->iface.get_prop(view, VIEW_PROP_APP_ID);
+	if (view->impl->get_prop) {
+		return view->impl->get_prop(view, VIEW_PROP_APP_ID);
 	}
 	return NULL;
 }
 
 const char *view_get_class(struct sway_view *view) {
-	if (view->iface.get_prop) {
-		return view->iface.get_prop(view, VIEW_PROP_CLASS);
+	if (view->impl->get_prop) {
+		return view->impl->get_prop(view, VIEW_PROP_CLASS);
 	}
 	return NULL;
 }
 
 const char *view_get_instance(struct sway_view *view) {
-	if (view->iface.get_prop) {
-		return view->iface.get_prop(view, VIEW_PROP_INSTANCE);
+	if (view->impl->get_prop) {
+		return view->impl->get_prop(view, VIEW_PROP_INSTANCE);
 	}
 	return NULL;
 }
 
+static void view_update_outputs(struct sway_view *view,
+		const struct wlr_box *before) {
+	struct wlr_output_layout *output_layout =
+		root_container.sway_root->output_layout;
+	struct wlr_box box = {
```

Comment:
Does this check need to be done for every surface that is contained in the view such as popups?